pasture-algorithms/src/normal_estimation.rs

```diff
@@ -0,0 +1,551 @@
+
+    if is_dense(point_cloud) {
+        point_count = point_cloud.len();
+        let mut diff_mean = vec![0.0; 4];
+        for point in point_cloud.iter_attribute::<Vector3<f64>>(&POSITION_3D) {
+            // calculate difference from the centroid for each point
+            diff_mean[0] = point.x - centroid[0];
+            diff_mean[1] = point.y - centroid[1];
+            diff_mean[2] = point.z - centroid[2];
+
+            covariance_matrix[(1, 1)] += diff_mean[1] * diff_mean[1];
+            covariance_matrix[(1, 2)] += diff_mean[1] * diff_mean[2];
+            covariance_matrix[(2, 2)] += diff_mean[2] * diff_mean[2];
+
+            let diff_x = diff_mean[0];
+            diff_mean.iter_mut().for_each(|x| *x = *x * diff_x);
+
+            covariance_matrix[(0, 0)] += diff_mean[0];
+            covariance_matrix[(0, 1)] += diff_mean[1];
+            covariance_matrix[(0, 2)] += diff_mean[2];
+        }
+    } else {
+        // in this case we dont know the number of points in the point cloud that are finite
+        for point in point_cloud.iter_attribute::<Vector3<f64>>(&POSITION_3D) {
+            if !is_finite(&point) {
+                continue;
+            }
+            // only compute the covariance matrix for finite points
+            let mut diff_mean = vec![0.0; 4];
+            // calculate difference from the centroid for each point
+            diff_mean[0] = point.x - centroid[0];
+            diff_mean[1] = point.y - centroid[1];
+            diff_mean[2] = point.z - centroid[2];
+
+            covariance_matrix[(1, 1)] += diff_mean[1] * diff_mean[1];
+            covariance_matrix[(1, 2)] += diff_mean[1] * diff_mean[2];
+            covariance_matrix[(2, 2)] += diff_mean[2] * diff_mean[2];
+
+            let diff_x = diff_mean[0];
+            diff_mean.iter_mut().for_each(|x| *x = *x * diff_x);
+
+            covariance_matrix[(0, 0)] += diff_mean[0];
+            covariance_matrix[(0, 1)] += diff_mean[1];
+            covariance_matrix[(0, 2)] += diff_mean[2];
+            point_count = point_count + 1;
+        }
+    }
+
+    covariance_matrix[(1, 0)] = covariance_matrix[(0, 1)];
+    covariance_matrix[(2, 0)] = covariance_matrix[(0, 2)];
+    covariance_matrix[(2, 1)] = covariance_matrix[(1, 2)];
+
+    return point_count;
+}
+
+/// find the eigen value solution if the highest degree of the polynomial is 2
+fn solve_polynomial_quadratic(
+    coefficient_2: &f64,
+    coefficient_1: &f64,
+    eigen_values: &mut Vec<f64>,
+) {
+    eigen_values[0] = 0.0;
+
+    let mut delta = coefficient_2 * coefficient_2 - 4.0 * coefficient_1;
+
+    if delta < 0.0 {
+        delta = 0.0;
+    }
+
+    let sqrt_delta = f64::sqrt(delta);
+
+    eigen_values[2] = 0.5 * (coefficient_2 + sqrt_delta);
+    eigen_values[1] = 0.5 * (coefficient_2 - sqrt_delta);
+}
+
+/// solve the polynomial to find the eigen values for a given covariance matrix
+fn solve_polynomial(covariance_matrix: &DMatrix<f64>, eigen_values: &mut Vec<f64>) {
```

Comment:
Is there nothing in `nalgebra` that does this? `nalgebra` has a lot of linear algebra stuff, so maybe there is something to handle covariance matrices as well?